kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/model/ingress/IngressAllocator.java

```diff
@@ -49,19 +63,66 @@ public static ProxyIngressModel allocateProxyIngressModel(KafkaProxy primary,
+                lastIdentifyingPort = ports.addAndGet(toAllocate) - 1;
             }
-            int firstIdentifyingPort = ports.get();
-            int lastIdentifyingPort = ports.addAndGet(toAllocate) - 1;
-            return new ProxyIngressModel.IngressModel(resource.createInstance(firstIdentifyingPort, lastIdentifyingPort), exception);
+            if (ingressDefinition.requiresSharedSniPort()) {
+                sharedSniPort = SHARED_SNI_PORT;
+            }
+
+            IngressModel ingressModel = ingressDefinition.createIngressModel(firstIdentifyingPort, lastIdentifyingPort, sharedSniPort);
+            return new ProxyIngressModel.IngressModelResult(ingressModel, exception);
         }).toList();
     }
+
+    static Stream<IngressDefinition> buildIngressDefinitions(KafkaProxy primary, ClusterResolutionResult clusterResult) {
+        VirtualKafkaCluster cluster = clusterResult.cluster();
+        return clusterResult.ingressResolutionResults().stream()
+                .flatMap(
+                        ingressResolutionResult -> {
+                            Optional<KafkaProxyIngress> maybeIngress = ingressResolutionResult.ingressResolutionResult().maybeReferentResource();
+                            if (maybeIngress.isPresent()) {
+                                Optional<KafkaService> maybeService = clusterResult.serviceResolutionResult().maybeReferentResource();
+                                // todo, maybe we should not include the case where the service does not resolve, it's optimistic to assume its using the default node id range
+                                List<NodeIdRanges> nodeIdRanges = maybeService.map(s -> s.getSpec().getNodeIdRanges()).orElse(DEFAULT_NODE_ID_RANGES);
+                                return Stream.of(buildIngressDefinition(primary, cluster, maybeIngress.get(), nodeIdRanges, ingressResolutionResult.ingress().getTls()));
+                            }
+                            else {
+                                // skip unresolved ingresses
+                                return Stream.empty();
+                            }
+                        });
+    }
+
+    private static IngressDefinition buildIngressDefinition(KafkaProxy primary,
+                                                            VirtualKafkaCluster cluster,
+                                                            KafkaProxyIngress ingress,
+                                                            List<NodeIdRanges> nodeIdRanges,
+                                                            @Nullable Tls tls) {
+        ClusterIP clusterIP = ingress.getSpec().getClusterIP();
+        LoadBalancer loadBalancer = ingress.getSpec().getLoadBalancer();
+        if (clusterIP != null) {
+            return new ClusterIPIngressDefinition(ingress, cluster, primary, nodeIdRanges, tls);
+        }
+        else if (loadBalancer != null) {
+            return new LoadBalancerDefinition(ingress, loadBalancer, tls);
```

Comment:
rough edge warning: this will blow up if `Tls` is null. LoadBalancer implicitly requires `Tls`. I _think_ it can't get to here in reality because the VKC reconciliation will detect the inconcistency and set a ResolvedRefs=false condition with `InvalidReferencedResource`.